src/ripple/rpc/impl/KeypairForSignature.cpp

```diff
@@ -18,81 +18,119 @@
+    char const* sigType = nullptr;
+    int count = 0;
+    for (auto t : sigTypes)
+    {
+        if (params.isMember (t))
+        {
+            ++count;
+            sigType = t;
+        }
+    }
 
-    int const count =
-        (hasPassphrase ? 1 : 0) +
-        (hasSeed ? 1 : 0) +
-        (hasHexSeed ? 1 : 0);
+    if (count != 1)
+    {
+        error = RPC::make_param_error (
+            "Exactly one of the following must be specified: " +
+            std::string(jss::passphrase) + ", " +
+            std::string(jss::seed) + " or " +
+            std::string(jss::seed_hex));
+        return boost::none;
+    }
 
-    if (count == 1)
+    // Make sure a string is present
+    if (! params[sigType].isString())
     {
-        if (hasSeed)
-            return parseBase58<Seed> (params[jss::seed].asString());
+        error = RPC::expected_field_error (sigType, "string");
+        return boost::none;
+    }
 
-        if (hasPassphrase)
-            return parseGenericSeed (params[jss::passphrase].asString());
+    auto const fieldContents = params[sigType].asString();
 
-        if (hasHexSeed)
-        {
-            uint128 seed;
+    // Convert string to seed.
+    boost::optional<Seed> seed;
 
-            if (seed.SetHexExact (params[jss::seed_hex].asString()))
-                return Seed { Slice(seed.data(), seed.size()) };
+    if (sigType == jss::seed.c_str())
+        seed = parseBase58<Seed> (fieldContents);
+    else if (sigType == jss::passphrase.c_str())
+        seed = parseGenericSeed (fieldContents);
+    else if (sigType == jss::seed_hex.c_str())
+    {
+        uint128 s;
 
-            return boost::none;
-        }
+        if (s.SetHexExact (fieldContents))
+            seed.emplace (Slice(s.data(), s.size()));
     }
 
-    return boost::none;
+    if (!seed)
+        error = rpcError (rpcBAD_SEED);
+
+    return seed;
 }
 
 std::pair<PublicKey, SecretKey>
 keypairForSignature (Json::Value const& params, Json::Value& error)
 {
     bool const has_key_type  = params.isMember (jss::key_type);
-    bool const hasPassphrase = params.isMember (jss::passphrase);
-    bool const hasSecret     = params.isMember (jss::secret);
-    bool const hasSeed       = params.isMember (jss::seed);
-    bool const hasSeedHex    = params.isMember (jss::seed_hex);
-
-    int const n_secrets =
-        (hasPassphrase ? 1 : 0) +
-        (hasSecret ? 1 : 0) +
-        (hasSeed ? 1 : 0) +
-        (hasSeedHex ? 1 : 0);
-
-    if (n_secrets == 0)
+
+    // All of the signature types we allow, but only one at a time.
+    // The array should be constexpr, but that makes Visual Studio unhappy.
+    static char const* const sigTypes[]
     {
-        error = RPC::missing_field_error (jss::secret);
-        return { };
+        jss::passphrase.c_str(),
+        jss::secret.c_str(),
+        jss::seed.c_str(),
+        jss::seed_hex.c_str()
+    };
+
+    // Identify which signature type is in use.
```

Comment:
I think "Identify which secret key type was provided" would be more appropriate here?